draft-ietf-ppm-dap.md

```diff
@@ -1267,106 +1334,173 @@ This message consists of:
-as described in {{agg-continue-flow}}. Otherwise, the Leader should abandon the
-aggregation job entirely.
+criteria in {{aggregation-job-validation}}. If the message is invalid, the
+Leader MUST abort the aggregation job.
+
+Otherwise, the Leader proceeds as follows with each report:
+
+1. If the inbound prep step has type "continue", then the Leader computes
+
+   ~~~
+   (state, outbound) = VDAF.ping_pong_req(agg_param, state, inbound)
+   ~~~
+
+   where `inbound` is the message payload. If `outbound != None`, then the
+   Leader stores `state` and `outbound` and proceeds to
+   {{aggregation-leader-continuation}}. If `outbound == None`, then the
+   preparation process is complete: either `state == Rejected()`, in which case
+   the Leader rejects the report and removes it from the candidate set; or
+   `state == Finished(out_share)`, in which case preparation is complete and the
+   Leader stores the output share for use in the collection sub-protocol
+   {{collect-flow}}.
+
+1. Else if the type is "rejected", then the Leader rejects the report and
+   removes it from the candidate set. Unless the error is `report_too_early`,
+   the Leader MUST NOT include the report in a subsequent aggregation job.
+   Otherwise, the Leader MAY include the report in a subsequent aggregation
+   job.
+
+1. Else the type is invalid, in which case the Leader MUST abort the
+   aggregation job.
+
+(Note: Since VDAF preparation completes in a constant number of rounds, it will
+never be the case that some reports are completed and others are not.)
 
 #### Helper Initialization {#aggregation-helper-init}
 
-Each Helper begins their portion of the aggregate initialization when they
-receive an AggregationJobInitReq message from the Leader. For each ReportShare
+The Helper begins an aggregation job when it receives an
+`AggregationJobInitReq` message from the Leader. For each `PrepareStep`
 conveyed by this message, the Helper attempts to initialize VDAF preparation
 (see {{Section 5.1 of !VDAF}}) just as the Leader does. If successful, it
-includes its prepare message in its response that the Leader will use to
-continue the process.
+includes the result in its response that the Leader will use to continue
+preparing the report.
 
 To begin this process, the Helper first checks if it recognizes the task ID. If
-not, then it MUST abort with error `unrecognizedTask`. Then the Helper checks
-that the report IDs in `AggregationJobInitReq.report_shares` are all distinct.
-If two ReportShare values have the same report ID, then the helper MUST abort
-with error `invalidMessage`. If this check succeeds, the helper then attempts to
-recover each input share in `AggregationJobInitReq.report_shares` as follows:
+not, then it MUST abort with error `unrecognizedTask`.
+
+Next, the Helper checks that the report IDs in
+`AggregationJobInitReq.prepare_steps` are all distinct. If two prep step
+messages have the same report ID, then the Helper MUST abort with error
```

Comment:
Technically it's the `report_share` inside the `PrepareStep` that has a report ID. I'm not sure if that's worth clarifying.